kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/VirtualKafkaClusterReconciler.java

```diff
@@ -42,18 +64,59 @@ public VirtualKafkaClusterReconciler(Clock clock) {
+                .collect(Collectors.toCollection(TreeSet::new));
+        missingProxies.removeAll(existingProxies);
+
+        var existingServices = context.getSecondaryResource(KafkaService.class, SERVICES_EVENT_SOURCE_NAME).stream()
+                .map(ResourcesUtil::toLocalRef)
+                .collect(Collectors.toSet());
+        TreeSet<LocalRef<KafkaService>> missingServices = Optional.ofNullable(cluster.getSpec())
+                .map(VirtualKafkaClusterSpec::getTargetKafkaServiceRef)
+                .stream()
+                .collect(Collectors.toCollection(TreeSet::new));
+        missingServices.removeAll(existingServices);
+
+        var existingIngresses = context.getSecondaryResourcesAsStream(KafkaProxyIngress.class)
+                .map(ResourcesUtil::toLocalRef)
+                .collect(Collectors.toSet());
+        TreeSet<LocalRef<KafkaProxyIngress>> missingIngresses = Optional.ofNullable(cluster.getSpec())
+                .map(VirtualKafkaClusterSpec::getIngressRefs)
+                .stream().flatMap(Collection::stream)
+                .collect(Collectors.toCollection(TreeSet::new));
+        missingIngresses.removeAll(existingIngresses);
+
+        var existingFilters = context.getSecondaryResourcesAsStream(KafkaProtocolFilter.class)
+                .map(ResourcesUtil::toLocalRef)
+                .collect(Collectors.toSet());
+        TreeSet<LocalRef<KafkaProtocolFilter>> missingFilters = Optional.ofNullable(cluster.getSpec())
+                .map(VirtualKafkaClusterSpec::getFilterRefs)
+                .stream().flatMap(Collection::stream)
+                .collect(Collectors.toCollection(TreeSet::new));
+        missingFilters.removeAll(existingFilters);
 
         Condition condition;
-        if (proxyOpt.isPresent()) {
+        if (missingProxies.isEmpty()
+                && missingServices.isEmpty()
+                && missingIngresses.isEmpty()
+                && missingFilters.isEmpty()) {
             condition = ResourcesUtil.newResolvedRefsTrue(clock, cluster);
         }
         else {
+            Stream<String> proxyMsg = getMissingProxy("spec.proxyRef references ", KafkaProxy.class, missingProxies);
+            Stream<String> serviceMsg = getMissingProxy("spec.targetKafkaServiceRef references ", KafkaService.class, missingServices);
+            Stream<String> ingressMsg = getMissingProxy("spec.ingressRefs references ", KafkaProxyIngress.class, missingIngresses);
+            Stream<String> filterMsg = getMissingProxy("spec.filterRefs references ", KafkaProtocolFilter.class, missingFilters);
+
             condition = ResourcesUtil.newResolvedRefsFalse(clock,
                     cluster,
-                    "spec.proxyRef.name",
-                    "KafkaProxy not found");
+                    "ReferencedResourcesNotFound",
+                    Stream.concat(Stream.concat(Stream.concat(proxyMsg, serviceMsg), ingressMsg), filterMsg).collect(Collectors.joining("; ")));
```

Comment:
We could extract our own varargs concat method. How about `Stream.of(proxyMsg, serviceMsg, ingressMsg, filterMsg).flatMap(identity())`?

Reply:
Thank you!